src/fs/pfs_filesystem_impl.cpp

```diff
@@ -0,0 +1,505 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * File Created: 2022-06-01
+ * Author: xuchaojie
+ */
+
+#include "src/fs/pfs_filesystem_impl.h"
+
+#include <glog/logging.h>
+#include <sys/vfs.h>
+#include <sys/utsname.h>
+#include <linux/version.h>
+#include <dirent.h>
+#include <brpc/server.h>
+#include <butil/iobuf.h>
+
+#include <pfs_api.h>
+#include <sys/uio.h>
+
+#include <string>
+#include <vector>
+
+#include "src/common/string_util.h"
+
+
+#define NVME_ALIGN 8
+#define SECTOR_SIZE 512
+
+namespace curve {
+namespace fs {
+
+std::string DumpIOVec(const struct iovec *vec, size_t nvec) {
+    std::stringstream ss;
+    ss << "{";
+    for (size_t i = 0; i < nvec; ++i) {
+        ss << "[" << uintptr_t(vec[i].iov_base) << ","
+           << vec[i].iov_len << "]";
+    }
+    ss << "}";
+    return ss.str();
+}
+
+ssize_t preadv_dispatch(int fd, const struct iovec *vector,
+                        int nvec, off_t offset) {
+    if (PFS_FD_ISVALID(fd)) {
+        int ret = pfs_preadv_dma(fd, vector, nvec, offset);
+        if (ret < 0) {
+            LOG(ERROR) << "preadv_dispatch failed, ret: " << ret
+                       << ", fd: " << fd
+                       << ", nvec: " << nvec
+                       << ", offset: " << offset
+                       << ", iovec: " << DumpIOVec(vector, nvec);
+        }
+        return ret;
+    }
+    return preadv(fd, vector, nvec, offset);
+}
+
+ssize_t pwritev_dispatch(int fd, const struct iovec *vector,
```

Comment:
Explain the function of this function

Reply:
It is the hook function used by iobuf to dispatch io use dma or not.
we add a comment here.